text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,848 @@
+```
+
+`~const` is derived from "approximately", meaning "conditionally" in this context, or specifically "const impl required if called in const context".
+It is the opposite of `?` (prexisting for `?Sized` bounds), which also means "conditionally", but from the other direction: `?const` (not proposed here, see the alternatives section for why it was rejected) would mean "no const impl required, even if called in const context".
+See [this alternatives section](#make-all-const-fn-arguments-const-trait-by-default-and-require-an-opt-out-const-trait) for an explanation of why we do not use a `?const` scheme.
+
+### Const fn
+
+`const` fn have always been and will stay "always const" functions.
+
+It may appear that a function is suddenly "not a const fn" if it gets passed a type that doesn't satisfy
+the constness of the corresponding trait bound. E.g.
+
+```rust
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self {
+        Foo
+    }
+}
+
+const fn bar<T: ~const Clone>(t: &T) -> T { t.clone() }
+const BAR: Foo = bar(Foo); // ERROR: `Foo`'s `Clone` impl is not for `const Clone`.
+```
+
+But `bar` is still a `const` fn and you can call it from a const context, it will just fail some trait bounds. This is no different from 
+
+```rust
+const fn dup<T: Copy>(a: T) -> (T, T) {(a, a)}
+const FOO: (String, String) = dup(String::new());
+```
+
+Here `dup` is always const fn, you'll just get a trait bound failure if the type you pass isn't `Copy`.
+
+This may seem like language lawyering, but that's how the impl works and how we should be talking about it.
+
+It's actually important for inference and method resolution in the nonconst world today.
+You first figure out which method you're calling, then you check its bounds.
+Otherwise it would at least seem like we'd have to allow some SFINAE or method overloading style things,
+which we definitely do not support and have historically rejected over and over again.
+
+
+### `~const Destruct` trait
+
+The `Destruct` trait enables dropping types within a const context.
+
+```rust
+const fn foo<T>(t: T) {
+    // `t` is dropped here, but we don't know if we can evaluate its `Drop` impl (or that of its fields' types)
+}
+const fn baz<T: Copy>(t: T) {
+    // Fine, `Copy` implies that no `Drop` impl exists
+}
+const fn bar<T: ~const Destruct>(t: T) {
+    // Fine, we can safely invoke the destructor of `T`.
+}
+```
+
+When a value of a generic type goes out of scope, it is dropped and (if it has one) its `Drop` impl gets invoked.
+This situation seems no different from other trait bounds, except that types can be dropped without implementing `Drop`
+(as they can contain types that implement `Drop`). In that case the type's drop glue is invoked.
+
+The `Destruct` trait is a bound for whether a type has drop glue. This is trivally true for all types.
+
+`~const Destruct` trait bounds are satsifed only if the type has a `const Drop` impl or all of the types of its components
+are `~const Destruct`.
```

Comment:
```suggestion
`~const Destruct` trait bounds are satisfied only if the type's `Drop` impl (if any) is `const` and all of the types of
its components are `~const Destruct`.
```

(This seems to be the currently implemented behavior and is what makes sense IMO: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0939069a19d8c571d88771e6f9e69713))